File: src/main/java/org/beatonma/lib/ui/activity/ActivityBuilder.java
package org.beatonma.lib.ui.activity;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import android.util.Pair;
import android.view.View;

import org.beatonma.lib.core.util.Sdk;
import org.beatonma.lib.log.Log;
import org.beatonma.lib.ui.activity.transition.BaseTransform;

import java.io.Serializable;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by Michael on 09/09/2016.
 * 
 * A builder to simplify initiation of non-trivial Activity instances
 * This is intended for subclasses of BaseActivity, but should work for any Activity
 *
 * Rationale: Static methods for simplifying the adding of extras and suchlike are
 * all well and good but can get messy if child classes need more and more specialised versions.
 * Using this builder we can support arbitrary argument sets in a clean manner.
 */

@SuppressWarnings("unused")
public class ActivityBuilder {
    protected final static String TAG = "ActivityBuilder";

    private Context mContext;
    private androidx.fragment.app.Fragment mFragment;
    private boolean mForResult = false;
    private int mRequestCode;
    private Intent mIntent;
    private List<Pair<View, String>> mSharedViews;
    private Bundle mExtras;

    private Class<? extends BaseTransform> mTransform;

    /*
     * If this activity is started by user interaction, identify the view which should be used
     * as the source for any animation
     */
    private View mAnimationSource;

    /*
     * If both the calling and target activities are instances of PopupActivity
     * automatically include the card and background overlay in shared view animation.
     */
    private boolean mUsePopupTransition = true;

    /**
     *
     */
    public static Constructor with(@NonNull final Context context) {
        return new Constructor(context);
    }

    public static ActivityBuilder forActivity(@NonNull final Context context,
                                              @NonNull final Class cls) {
        return new ActivityBuilder(context, cls);
    }

    public static ActivityBuilder forActivity(@NonNull final Context context,
                                              @NonNull final Intent intent) {
        return new ActivityBuilder(context, intent);
    }

    private ActivityBuilder(@NonNull final Context context) {
        mContext = context;
        mSharedViews = new ArrayList<>();
        mExtras = new Bundle();
    }

    public ActivityBuilder(@NonNull final Context context, @NonNull final Class cls) {
        this(context, new Intent(context, cls));
    }

    public ActivityBuilder(@NonNull final Context context, @NonNull final Intent intent) {
        this(context);
        mIntent = intent;
    }

    public ActivityBuilder setClass(final Class cls) {
        mIntent = new Intent(mContext, cls);
        return this;
    }

    public ActivityBuilder setIntent(final Intent intent) {
        mIntent = intent;
        return this;
    }

    public ActivityBuilder forResult(final int requestCode) {
        mForResult = true;
        mRequestCode = requestCode;
        return this;
    }

    public ActivityBuilder forResult(final Fragment fragment, final int requestCode) {
        mFragment = fragment;
        mForResult = true;
        mRequestCode = requestCode;
        return this;
    }

    public ActivityBuilder addSharedViews(final Pair<View, String>... sharedViews) {
        Collections.addAll(mSharedViews, sharedViews);
        return this;
    }

    public ActivityBuilder addSharedView(final Pair<View, String> sharedView) {
        mSharedViews.add(sharedView);
        return this;
    }

    public ActivityBuilder addSharedView(final View view, final String transitionName) {
        mSharedViews.add(new Pair<> (view, transitionName));
        return this;
    }

    public ActivityBuilder addSharedView(final View view, final int transitionResourceId) {
        return addSharedView(view, mContext.getString(transitionResourceId));
    }

    public ActivityBuilder putExtra(final String name, final boolean value) {
        mExtras.putBoolean(name, value);
        return this;
    }

    public ActivityBuilder putExtra(final String name, final int value) {
        mExtras.putInt(name, value);
        return this;
    }

    public ActivityBuilder putExtra(final String name, final float value) {
        mExtras.putFloat(name, value);
        return this;
    }

    public ActivityBuilder putExtra(final String name, final String value) {
        mExtras.putString(name, value);
        return this;
    }

    public ActivityBuilder putExtra(final String name, final Serializable value) {
        mExtras.putSerializable(name, value);
        return this;
    }

    public ActivityBuilder putExtra(final String name, final Bundle value) {
        mExtras.putBundle(name, value);
        return this;
    }

    public ActivityBuilder putExtraStringArrayList(final String name, final ArrayList<String> value) {
        mExtras.putStringArrayList(name, value);
        return this;
    }

    public ActivityBuilder putExtraIntArrayList(final String name, final ArrayList<Integer> value) {
        mExtras.putIntegerArrayList(name, value);
        return this;
    }

    public ActivityBuilder putExtraParcelableArrayList(final String name, final ArrayList<? extends Parcelable> value) {
        mExtras.putParcelableArrayList(name, value);
        return this;
    }

    private ActivityBuilder putExtra(final String name, final Object value) throws ClassCastException {
        if (value instanceof Integer) {
            return putExtra(name, (Integer) value);
        }
        else if (value instanceof String) {
            return putExtra(name, (String) value);
        }
        else if (value instanceof Boolean) {
            return putExtra(name, (Boolean) value);
        }
        else if (value instanceof Float) {
            return putExtra(name, (Float) value);
        }
        else if (value instanceof ArrayList) {
            return putExtra(name, (ArrayList<String>) value);
        }
        else if (value instanceof Bundle) {
            return putExtra(name, (Bundle) value);
        }
        else if (value instanceof Serializable) {
            return putExtra(name, (Serializable) value);
        }
        return this;
    }

    /**
     * Copy the extras from another intent
     * @param intent
     * @return
     */
    public ActivityBuilder putExtras(final Intent intent) {
        mIntent.putExtras(intent);
        return this;
    }

    public ActivityBuilder putExtras(final Bundle bundle) {
        mIntent.putExtras(bundle);
        return this;
    }

    public View animationSource() {
        return mAnimationSource;
    }

    public ActivityBuilder animationSource(final View animationSource) {
        mAnimationSource = animationSource;
        return this;
    }

    public boolean usePopupTransition() {
        return mUsePopupTransition;
    }

    public ActivityBuilder usePopupTransition(final boolean popupTransition) {
        mUsePopupTransition = popupTransition;
        return this;
    }

    public ActivityBuilder transform(final Class<? extends BaseTransform> transform) {
        mTransform = transform;
        return this;
    }

    /**
     * Add extras to mIntent for the given transformation.
     * mTransform must implement the same addExtras(Intent, View) static method
     * as seen in BaseTransform.
     */
    private void addTransitionExtras() {
        if (mTransform != null) {
            try {
                Method addExtras = mTransform.getMethod("addExtras", Intent.class, View.class);
                addExtras.invoke(null, mIntent, mAnimationSource);
                return;
            }
            catch (NoSuchMethodException e) {
                Log.d(TAG, "addTransitionExtras(): Method not found: %s", e.toString());
            }
            catch (InvocationTargetException e) {
                Log.d(TAG, "addTransitionExtras(): Invocation error: %s %s",
                        e.getCause().toString(), e.getTargetException().toString());
            }
            catch (IllegalAccessException e) {
                Log.d(TAG, "addTransitionExtras(): Illegal access error: %s", e.toString());
            }
            catch (Exception e) {
                Log.d(TAG, "addTransitionExtras(): %s", e.toString());
            }
        }
        BaseTransform.addExtras(mIntent, mAnimationSource);
    }

    public void start() {
        if (mIntent == null) {
            throw new NullPointerException("Intent is missing. A target Class or Intent must be provided.");
        }

        mIntent.putExtras(mExtras);

        if (mContext instanceof BaseActivity) {
            startWithActivityContext();
        }
        else {
            mIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            mContext.startActivity(mIntent);
        }
    }

    @SuppressWarnings("NewApi")
    private void startWithActivityContext() {
        final Activity activity = (Activity) mContext;

        if (mUsePopupTransition) {
            if (mContext instanceof BaseActivity) {
                addSharedViews(((BaseActivity) mContext).getSharedViews());
                if (mContext instanceof PopupActivity) {
                    mIntent.putExtra(PopupActivity.EXTRA_CALLED_FROM_POPUP, true);
                }
            }
        }

        if (mAnimationSource != null) {
            if (Sdk.isKitkat()) {
//                BaseTransform.addExtras(mIntent, mAnimationSource);
                addTransitionExtras();
            }
            if (mSharedViews == null) {
                Log.w(TAG, "An animation source has been defined but no shared views. A shared view must be defined for this type of transition to work.");
            }
        }

        if (Sdk.isLollipop() && mSharedViews != null) {
            final Pair<View, String>[] sharedViews = new Pair[mSharedViews.size()];
            for (int i = 0; i < mSharedViews.size(); i++) {
                sharedViews[i] = mSharedViews.get(i);
            }
            Log.d(TAG, "shared views: %d", mSharedViews.size());
            final ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(activity, sharedViews);
            if (mForResult) {
                if (mFragment != null) {
                    mFragment.startActivityForResult(mIntent, mRequestCode, options.toBundle());
                }
                else {
                    Log.d(TAG, "Starting activity with animation");
                    activity.startActivityForResult(mIntent, mRequestCode, options.toBundle());
                }
            }
            else {
                Log.d(TAG, "Starting activity with animation");
                activity.startActivity(mIntent, options.toBundle());
            }
        }
        else {
            if (mForResult) {
                if (mFragment != null) {
                    mFragment.startActivityForResult(mIntent, mRequestCode);
                }
                else {
                    activity.startActivityForResult(mIntent, mRequestCode);
                }
            }
            else {
                activity.startActivity(mIntent);
            }
        }
    }


    public static class Constructor {
        private Context context;

        public Constructor(Context context) {
            this.context = context;
        }

        public ActivityBuilder target(Class cls) {
            return new ActivityBuilder(this.context, cls);
        }

        public ActivityBuilder target(Intent intent) {
            return new ActivityBuilder(this.context, intent);
        }
    }
}

File: src/main/java/org/beatonma/lib/ui/activity/ActivityBuilder.kt
package org.beatonma.lib.ui.activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.annotation.StringRes
import androidx.core.app.ActivityOptionsCompat
import androidx.core.util.Pair
import androidx.fragment.app.Fragment
import org.beatonma.lib.ui.activity.popup.BasePopupActivity
import org.beatonma.lib.ui.activity.popup.EXTRA_CALLED_FROM_POPUP
import org.beatonma.lib.ui.activity.transition.BaseTransform
import org.beatonma.lib.util.Sdk
import org.beatonma.lib.util.kotlin.extensions.autotag
import org.beatonma.lib.util.kotlin.extensions.stringCompat
import java.io.Serializable
import java.lang.ref.WeakReference
import kotlin.reflect.KClass

/**
 * Created by Michael on 09/09/2016. Rewritten for Kotlin 14/06/2018.
 *
 * A builder to simplify initiation of non-trivial Activity instances
 * This is intended for subclasses of BaseActivity, but should work for any Activity
 */
@SuppressWarnings("unused")
class ActivityBuilder(
        context: Any,   // Any Context instance, an attached Fragment, or an attached View
        cls: Class<*>,
        var requestCode: Int? = null,
        fragment: Fragment? = null,
        var animationSource: View? = null
) {
    private val weakContext: WeakReference<Context>
    private val weakFragment: WeakReference<Fragment>?

    private val intent: Intent
    val extras = Bundle()

    val sharedViews = mutableListOf<SharedView>()

    var usePopupAnimation: Boolean = true

    var transform: Class<out BaseTransform>? = null

    constructor(context: Any, cls: KClass<*>): this(context, cls.java)

    init {
        // Get a concrete Context object from the given source object (Context, Fragment, View)
        val strongContext: Context = when (context) {
            is Context -> context
            is Fragment -> {
                context.context ?: throw IllegalArgumentException(
                        "Fragment is not attached - unable to get Context")
            }
            // If a View is given as context it is set as the animation source
            // (unless a source is already defined)
            is View -> {
                animationSource = animationSource ?: context
                context.context ?: throw IllegalArgumentException(
                        "View is not attached - unable to get Context")
            }
            else -> throw IllegalArgumentException(
                    "context must be a Context instance, or an attached Fragment or View")
        }
        weakContext = WeakReference(strongContext)

        // Try to get a concrete Fragment object from either context or fragment parameters.
        val strongFragment = fragment ?: context as? Fragment
        weakFragment = if (strongFragment == null) null else WeakReference(strongFragment)

        intent = Intent(strongContext, cls)
    }

    fun addSharedView(view: View?, transitionName: String) {
        view ?: return
        sharedViews.add(SharedView(view, transitionName))
    }

    fun addSharedView(view: View?, @StringRes transitionNameResId: Int) {
        view ?: return
        weakContext.get()?.stringCompat(transitionNameResId)?.let {
            sharedViews.add(SharedView(view, it))
        }
    }

    fun addSharedViews(vararg sharedViews: SharedView) {
        this.sharedViews.addAll(sharedViews)
    }

    fun putExtra(key: String, value: Any) {
        when (value) {
            is Boolean -> extras.putBoolean(key, value)
            is Bundle -> extras.putBundle(key, value)
            is Int -> extras.putInt(key, value)
            is Float -> extras.putFloat(key, value)
            is Long -> extras.putLong(key, value)
            is String -> extras.putString(key, value)
            is Serializable -> extras.putSerializable(key, value)
        }
    }

    fun putExtras(bundle: Bundle) {
        extras.putAll(bundle)
    }

    fun putExtras(intent: Intent) {
        extras.putAll(intent.extras)
    }

    /**
     * Convenience for builder.apply {
     *     ...
     *     start()
     * }
     */
    inline fun startWith(block: (ActivityBuilder.() -> Unit)) {
        this.block()
        start()
    }

    fun start() {
        intent.putExtras(extras)
        val context = weakContext.get()

        if (context is BaseActivity) {
            startWithBaseActivityContext(context)
        } else {
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            context?.startActivity(intent)
        }
    }

    private fun startWithBaseActivityContext(activity: BaseActivity) {
        if (usePopupAnimation) {
            activity.sharedViews?.let { addSharedViews(*it) }
            if (activity is BasePopupActivity) {
                intent.putExtra(EXTRA_CALLED_FROM_POPUP, true)
            }
        }

        if (animationSource != null) {
            if (Sdk.isKitkat) addTransitionExtras()
            if (sharedViews.isEmpty()) {
                Log.w(autotag,
                        "An animation source has been defined but there are no shared views.")
            }
        }

        val sharedViewsBundle = createSharedViewsBundle(activity)
        val fragment = weakFragment?.get()
        if (fragment != null) fragment.startActivity(intent, requestCode, sharedViewsBundle)
        else activity.startActivity(intent, requestCode, sharedViewsBundle)
    }

    private fun createSharedViewsBundle(activity: BaseActivity): Bundle? {
        val sharedViews = Array(this.sharedViews.size) { sharedViews[it].asPair }

        return ActivityOptionsCompat.makeSceneTransitionAnimation(
                activity, *sharedViews).toBundle()
    }

    private fun addTransitionExtras() {
        transform?.let { t ->
            try {
                t.getMethod(
                        "addExtras", Intent::class.java, View::class.java)
                        ?.invoke(null, intent, animationSource)
            } catch (e: Exception) {
                // This doesn't matter too much - just an animation error
                Log.d(autotag, "addTransitionExtras() failed: $e")
            }
        }
        BaseTransform.addExtras(intent, animationSource)
    }
}

data class SharedView(val view: View, val transitionName: String) {
    val asPair: Pair<View, String> by lazy { Pair(view, transitionName) }
}

private fun Fragment.startActivity(
        intent: Intent,
        requestCode: Int? = null,
        optionsCompatBundle: Bundle? = null
) {
    if (requestCode != null) {
        if (Sdk.isLollipop) startActivityForResult(intent, requestCode, optionsCompatBundle)
        else startActivityForResult(intent, requestCode)
    } else {
        if (Sdk.isLollipop) startActivity(intent, optionsCompatBundle)
        else startActivity(intent)
    }
}

private fun Activity.startActivity(
        intent: Intent,
        requestCode: Int? = null,
        optionsCompatBundle: Bundle? = null
) {
    if (requestCode != null) {
        if (Sdk.isLollipop) startActivityForResult(intent, requestCode, optionsCompatBundle)
        else startActivityForResult(intent, requestCode)
    } else {
        if (Sdk.isLollipop) startActivity(intent, optionsCompatBundle)
        else startActivity(intent)
    }
}

/**
 * Convenience for ActivityBuilder(Context, KClass).startWith{ ... }
 */
inline fun Context.startActivity(cls: KClass<*>, block: ActivityBuilder.() -> Unit = {}) {
    ActivityBuilder(this, cls).startWith(block)
}

/**
 * Convenience for ActivityBuilder(Context, KClass).startWith{ ... }
 */
inline fun Fragment.startActivity(cls: KClass<*>, block: ActivityBuilder.() -> Unit = {}) {
    ActivityBuilder(this, cls).startWith(block)
}

/**
 * Convenience for ActivityBuilder(Context, KClass).startWith{ ... }
 */
inline fun View.startActivity(cls: KClass<*>, block: ActivityBuilder.() -> Unit = {}) {
    ActivityBuilder(this, cls).startWith(block)
}
